libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m

```diff
@@ -549,677 +293,275 @@ - (void)migrateUserDefaults {
-            if (newAccount.communities) {
-                newAccount.communities = [self.currentUser.communities arrayByAddingObject:communityData];
+        if (![currentAccount communityWithId:credentials.communityId]) {
+            if (currentAccount.communities) {
+                currentAccount.communities = [currentAccount.communities arrayByAddingObject:communityData];
             } else {
-                newAccount.communities = @[communityData];
+                currentAccount.communities = @[communityData];
             }
         }
-        [self setActiveCommunityId:newAccount.communityId];
+        [self setCurrentCommunityId:currentAccount.communityId];
     }
-    
-    // If our default user identity is currently the temporary user identity,
-    // we need to update it with the latest known good user identity.
-    if ([self.activeUserIdentity isEqual:self.temporaryUserIdentity]) {
-        [self log:SFLogLevelDebug format:@"Replacing temp user identity with %@", newAccount];
-        [self replaceOldUser:self.temporaryUserIdentity withUser:newAccount];
-    }
-    self.currentUser = newAccount;
+
+    [self saveAccountForUser:currentAccount error:nil];
     [self userChanged:change];
+    return currentAccount;
+}
+
+- (SFUserAccount*) currentUser {
+    if (!_currentUser) {
+        [_accountsLock lock];
+        NSData *resultData = nil;
+        NSUserDefaults *userDefaults = [NSUserDefaults msdkUserDefaults];
+        resultData = [userDefaults objectForKey:kUserDefaultsLastUserIdentityKey];
+        if (resultData) {
+             SFUserAccountIdentity *result = nil;
+            @try {
+                NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingWithData:resultData];
+                result = [unarchiver decodeObjectForKey:kUserDefaultsLastUserIdentityKey];
+                [unarchiver finishDecoding];
+                if(result)
+                   _currentUser = [self userAccountForUserIdentity:result];
+                else
+                    [self log:SFLogLevelError msg:@"Located current user Identity in NSUserDefaults but was not found in list of accounts managed by SFUserAccountManager."];
+            }
+            @catch (NSException *exception) {
+                [self log:SFLogLevelWarning msg:@"Could not parse current user identity from user defaults.  Setting to nil."];
+            }
+        }else {
+            [self log:SFLogLevelWarning msg:@"Could not parse locate user identity object from user defaults.  Setting to nil."];
+        }
+        [_accountsLock unlock];
+    }
+    return _currentUser;
 }
 
-- (void)applyIdData:(SFIdentityData *)idData {
-    self.currentUser.idData = idData;
+- (void)setCurrentUser:(SFUserAccount*)user {
+
+    BOOL userChanged = NO;
+    if(user!=_currentUser) {
+        [_accountsLock lock];
+        if (!user) {
+            //clear current user if  nil
+            _currentUser = nil;
```

Comment:
Technically, you want the `[self willChangeValueForKey:@"currentUser"]` / `didChange` dance anywhere in here that `_currentUser` is going to be changed.